src/order/locally_finite.lean

```diff
@@ -482,6 +482,192 @@ locally_finite_order.of_Icc' (α × β)
+    | (a : α),       ⊤, (x : α) := by simp [with_top.locally_finite_order._match_1, coe_eq_coe]
+    | (a : α), (b : α),       ⊤ := by simp [with_top.locally_finite_order._match_1]
+    | (a : α), (b : α), (x : α) := by simp [with_top.locally_finite_order._match_1, coe_eq_coe]
+    end,
+  finset_mem_Ico := λ a b x, match a, b, x with
+    |       ⊤,       b,       x := iff_of_false (not_mem_empty _)
+                                     (λ h, not_top_lt $ h.1.trans_lt h.2)
+    | (a : α),       ⊤,       ⊤ := by simp [with_top.locally_finite_order._match_3]
+    | (a : α),       ⊤, (x : α) := by simp [with_top.locally_finite_order._match_3, coe_eq_coe,
+                                        coe_lt_top]
+    | (a : α), (b : α),       ⊤ := by simp [with_top.locally_finite_order._match_3]
+    | (a : α), (b : α), (x : α) := by simp [with_top.locally_finite_order._match_3, coe_eq_coe,
+                                        coe_lt_coe]
+    end,
+  finset_mem_Ioc := λ a b x, match a, b, x with
+    |       ⊤,       b,       x := iff_of_false (not_mem_empty _)
+                                     (λ h, not_top_lt $ h.1.trans_le h.2)
+    | (a : α),       ⊤,       ⊤ := by simp [with_top.locally_finite_order._match_4, coe_lt_top]
+    | (a : α),       ⊤, (x : α) := by simp [with_top.locally_finite_order._match_4, coe_eq_coe,
+                                        coe_lt_coe]
+    | (a : α), (b : α),       ⊤ := by simp [with_top.locally_finite_order._match_4]
+    | (a : α), (b : α), (x : α) := by simp [with_top.locally_finite_order._match_4, coe_eq_coe,
+                                        coe_lt_coe]
+    end,
+  finset_mem_Ioo := λ a b x, match a, b, x with
+    |       ⊤,       b,       x := iff_of_false (not_mem_empty _)
+                                     (λ h, not_top_lt $ h.1.trans h.2)
+    | (a : α),       ⊤,       ⊤ := by simp [with_top.locally_finite_order._match_6, coe_lt_top]
+    | (a : α),       ⊤, (x : α) := by simp [with_top.locally_finite_order._match_6, coe_eq_coe,
+                                        coe_lt_coe, coe_lt_top]
+    | (a : α), (b : α),       ⊤ := by simp [with_top.locally_finite_order._match_6]
+    | (a : α), (b : α), (x : α) := by simp [with_top.locally_finite_order._match_6, coe_eq_coe,
+                                        coe_lt_coe]
+    end }
+
+variables (a b : α)
+
+@[simp] lemma Icc_coe_top :
+  Icc (a : with_top α) ⊤ = cons ⊤ ((Ici a).map embedding.coe_with_top)
+    (λ h, let ⟨x, _, hx⟩ := mem_map.1 h in coe_ne_top hx) := rfl
+
+@[simp] lemma Icc_coe_coe : Icc (a : with_top α) b = (Icc a b).map embedding.coe_with_top := rfl
+
+@[simp] lemma Ico_coe_top : Ico (a : with_top α) ⊤ = (Ici a).map embedding.coe_with_top := rfl
+@[simp] lemma Ico_coe_coe : Ico (a : with_top α) b = (Ico a b).map embedding.coe_with_top := rfl
+
+@[simp] lemma Ioc_coe_top :
+  Ioc (a : with_top α) ⊤ = cons ⊤ ((Ioi a).map embedding.coe_with_top)
+    (λ h, let ⟨x, _, hx⟩ := mem_map.1 h in coe_ne_top hx) := rfl
```

Comment:
Those `cons ⊤` are defeq to `insert_none`. That is, this works:
```suggestion
lemma Ioc_coe_top : Ioc (a : with_top α) ⊤ = insert_none (Ioi a) := rfl
```
Do you want me to use `insert_none`/define `insert_top` and `insert_bot`?

Reply:
That looks a lot nicer. 
Not sure about (ab)using the definitional equality between `option`/`with_top`/`with_bot`. I *think* we allow ourselves to do this, so that we don't have to redefine everything for `with_top`/`with_bot`?
So I'd say use `insert_none`.

Reply:
Note that changing it in the definitions breaks the `simp`s because of bad definitional unfolding, as Eric already pointed out.

Reply:
Are you talking about the `embedding` stuff? That is something different. `embedding.some` is defined using the `some` function, while you want the version that is defined using `coe` (which is definitionally but not syntactically equal). Really, `embedding.some` should never be used or replaced by a version that has `coe` as its underlying function.

Reply:
I mean that replacing
```
  finset_Ioc := λ a b, match a, b with
    |      ⊤,      _ := ∅
    | (a : α),      ⊤ := cons (⊤ : with_top α) ((Ioi a).map embedding.coe_with_top)
                              (λ h, let ⟨x, _, hx⟩ := mem_map.1 h in coe_ne_top hx)
    | (a : α), (b : α) := (Ioc a b).map embedding.coe_with_top
    end,
```
by
```
  finset_Ioc := λ a b, match a, b with
    |      ⊤,      _ := ∅
    | (a : α),      ⊤ := insert_none (Ioi a)
    | (a : α), (b : α) := (Ioc a b).map embedding.coe_with_top
    end,
```
breaks the `simp`s in ``finset_mem_Ioc`.
So why should we not use  `embedding.coe_with_top`/`embedding.coe_with_bot`? They do solve a problem here. Without them, `simp` gets lost because of the leaky implementation of `with_top`/`with_bot`.

Reply:
> So why should we not use embedding.coe_with_top/embedding.coe_with_bot? They do solve a problem here. Without them, simp gets lost because of the leaky implementation of with_top/with_bot.

This is false. As I said in my last message, the `option` vs `with_top`/`with_bot` doesn't matter. What matters is `some` vs `coe`. I just tested, and replaced all occurrences of `embedding.coe_with_[top|bot]` by the following, and it compiles fine:
```lean
@[simps { fully_applied := ff }] protected def embedding.option_coe {α} : α ↪ option α :=
⟨coe, option.some_injective α⟩
```
This is what I mean with `embedding.some` is bad and should be replaced by `embedding.option_coe`.

The issue with `insert_none` is the same: some `simp` lemmas use `some` and some use `coe`. If I add this lemma above your definition, then it compiles fine with the `insert_none`'s.
It would be great it we reformulated `option.mem_def` so that it uses the coercion.
```lean
@[simp] theorem my_mem_def {a : α} {b : option α} : a ∈ b ↔ b = a :=
iff.rfl
```

Reply:
Maybe in this PR you can add the alternatives `embedding.option_coe` and `my_mem_def` (maybe renamed to something like `option.mem_iff_eq_coe`?) without the `simp` attribute.
In a follow-up PR you could make the new `mem_def` lemma the default `simp` lemma (if you want); that is likely going to break stuff elsewhere.

Reply:
Okay yeah indeed it works. Thanks!